bst.py

```diff
@@ -0,0 +1,129 @@
+            yield "\t%s -> null%s;" % (self.val, r)
+        if self.right is not None:
+            yield "\t%s -> %s;" % (self.val, self.right.val)
+            for i in self.right._get_dot():
+                yield i
+        elif self.left is not None:
+            r = random.randint(0, 1e9)
+            yield "\tnull%s [shape=point];" % r
+            yield "\t%s -> null%s;" % (self.val, r)
+
+
+class BinarySearchTree(object):
+    root = None
+
+    def __init__(self):
+        """Initialize a binary search tree"""
+        self.root = None
+        self.set = set()
+
+    def insert(self, val):
+        """Insert a node, checks if duplicate"""
+        if val not in self.set:
+            if not self.root:
+                self.root = Node(val)
+            else:
+                current = self.root
+                while current:
+                    if val < current.val:
+                        if current.left is None:
+                            current.left = Node(val, parent=current)
+                            break
+                        current = current.left
+                    else:
+                        if current.right is None:
+                            current.right = Node(val, parent=current)
+                            break
+                        current = current.right
+            self.set.add(val)
+
+    def contains(self, val):
+        """Return a True/False if node in the tree"""
+        return val in self.set
+
+    def size(self):
+        """Return number of items in  tree"""
+        return len(self.set)
+
+    def depth(self):
+        """Return number of levels in tree"""
+        if self.root:
+            return self.root.depth()
+
+    def balance(self):
+        """If more values on right returns 1, if left -1. Else 0"""
+        if self.depth == 1:
+            return 0
+        else:
+            return self.root.balance()
+
+    def get_dot(self):
+        """return the tree with root 'self' as a dot graph for visualization"""
+        return "digraph G{\n%s}" % ("" if self.root is None else (
+            "\t%s;\n%s\n" % (
+                self.root.val,
+                "\n".join(self.root._get_dot())
+            )
+        ))
+
+
+if __name__ == '__main__':
+    # Create a random Binary Search Tree.
+    # Call: dot -Tpng test.gv -o testGraph.png
+    # from cmdline to make a png file containing a visual
+    # representation of the tree.
+    tree = BinarySearchTree()
+    for i in range(10):
+        tree.insert(random.randint(1, 100))
+    dot_tree = tree.get_dot()
+    with open('test.gv', 'w') as fh:
+        fh.write(dot_tree)
+
+    tree2 = BinarySearchTree()
+    for i in range(100):
+        tree2.insert(i)
+
+    # worst case
+    t = timeit.Timer('tree2.contains(99)', 'from __main__ import tree2')
+    print t.timeit()
+
+    # best case
+    t = timeit.Timer('tree2.contains(1)', 'from __main__ import tree2')
+    print t.timeit()
```

Comment:
Not really a good test here.  Best case should be finding the value 99 in a well-balanced tree.